diff --git a/inc/common/oscore_edhoc_error.h b/inc/common/oscore_edhoc_error.h
index 244e9c2..f0e06c3 100644
--- a/inc/common/oscore_edhoc_error.h
+++ b/inc/common/oscore_edhoc_error.h
@@ -79,6 +79,10 @@ enum err {
 	oscore_interaction_duplicated_token = 221,
 	oscore_interaction_not_found = 222,
 	oscore_wrong_uri_path = 223,
+
+	/* SCHC errors */
+	schc_inner_comp_error = 300,
+	schc_inner_decomp_error = 301
 };
 
 /*This macro checks if a function returns an error and if so it propagates 
diff --git a/inc/oscore.h b/inc/oscore.h
index f3b7e3d..745c654 100644
--- a/inc/oscore.h
+++ b/inc/oscore.h
@@ -23,6 +23,8 @@
 #include "common/oscore_edhoc_error.h"
 #include "common/print_util.h"
 
+#include "oscore_ext.h"
+
 /*
  * When OSCORE is used with fixed keys, i.e., no re-keying with EDHOC 
  * after reboot the SSN needs to be stored at runtime in NVM and restored 
@@ -114,7 +116,7 @@ enum err oscore_context_init(struct oscore_init_params *params,
  * @return	err
  */
 enum err oscore2coap(uint8_t *buf_in, uint32_t buf_in_len, uint8_t *buf_out,
-		     uint32_t *buf_out_len, struct context *c);
+		     uint32_t *buf_out_len, struct context *c, plaintext_cb_t oscore_msg_inner_decompression);
 
 /**
  *@brief 	Converts a CoAP packet to OSCORE packet
@@ -128,6 +130,6 @@ enum err oscore2coap(uint8_t *buf_in, uint32_t buf_in_len, uint8_t *buf_out,
  */
 enum err coap2oscore(uint8_t *buf_o_coap, uint32_t buf_o_coap_len,
 		     uint8_t *buf_oscore, uint32_t *buf_oscore_len,
-		     struct context *c);
+		     struct context *c, plaintext_cb_t oscore_msg_inner_compression);
 
 #endif
diff --git a/inc/oscore_ext.h b/inc/oscore_ext.h
index e69de29..7b08595 100644
--- a/inc/oscore_ext.h
+++ b/inc/oscore_ext.h
@@ -0,0 +1,16 @@
+#ifndef OSCORE_EXT_H
+#define OSCORE_EXT_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+/** Callback used to process inner Compression and decompression
+ * plaintext: oscore plaintext
+ * plaintext_size: oscore plaintext size
+ *
+ * Returned value;
+ * - bool: True if success otherwise it will return False
+ **/
+typedef bool (*plaintext_cb_t)(uint8_t *plaintext, uint32_t *plaintext_size);
+
+#endif
\ No newline at end of file
diff --git a/makefile_config.mk b/makefile_config.mk
index 0bc6951..afd08d6 100644
--- a/makefile_config.mk
+++ b/makefile_config.mk
@@ -158,5 +158,5 @@ FEATURES += -DI_OPTIONS_BUFF_MAX_LEN=100
 
 
 #CRYPTO_ENGINE += -DTINYCRYPT
-CRYPTO_ENGINE += -DCOMPACT25519
+#CRYPTO_ENGINE += -DCOMPACT25519
 CRYPTO_ENGINE += -DMBEDTLS
diff --git a/src/oscore/coap2oscore.c b/src/oscore/coap2oscore.c
index 1ef588b..22aa552 100644
--- a/src/oscore/coap2oscore.c
+++ b/src/oscore/coap2oscore.c
@@ -589,7 +589,7 @@ static enum err encrypt_wrapper(struct byte_array *plaintext,
  */
 enum err coap2oscore(uint8_t *buf_o_coap, uint32_t buf_o_coap_len,
 		     uint8_t *buf_oscore, uint32_t *buf_oscore_len,
-		     struct context *c)
+		     struct context *c, plaintext_cb_t oscore_msg_inner_compression)
 {
 	struct o_coap_packet o_coap_pkt;
 	struct byte_array buf;
@@ -644,6 +644,12 @@ enum err coap2oscore(uint8_t *buf_o_coap, uint32_t buf_o_coap_len,
 	/* Combine code, E-options and payload of CoAP to plaintext */
 	TRY(plaintext_setup(&o_coap_pkt, e_options, e_options_cnt, &plaintext));
 
+	//if oscore_msg_inner_compression is given to our func we enable Inner compression plaintext here
+    if (oscore_msg_inner_compression != NULL) {
+        if (!oscore_msg_inner_compression(plaintext.ptr, &(plaintext.len)))
+            return schc_inner_comp_error;
+    }
+
 	/* Generate ciphertext array */
 	BYTE_ARRAY_NEW(ciphertext, MAX_CIPHERTEXT_LEN,
 		       plaintext.len + AUTH_TAG_LEN);
diff --git a/src/oscore/oscore2coap.c b/src/oscore/oscore2coap.c
index 97a07bc..fae8dc9 100644
--- a/src/oscore/oscore2coap.c
+++ b/src/oscore/oscore2coap.c
@@ -255,7 +255,8 @@ decrypt_wrapper(struct byte_array *ciphertext, struct byte_array *plaintext,
 		struct context *c,
 		struct compressed_oscore_option *new_nonce_oscore_option,
 		struct o_coap_packet *input_oscore,
-		struct o_coap_packet *output_coap)
+		struct o_coap_packet *output_coap,
+		plaintext_cb_t oscore_msg_inner_decompression)
 {
 	BYTE_ARRAY_NEW(new_nonce, NONCE_LEN, NONCE_LEN);
 	struct byte_array nonce;
@@ -301,6 +302,11 @@ decrypt_wrapper(struct byte_array *ciphertext, struct byte_array *plaintext,
 	TRY(oscore_cose_decrypt(ciphertext, plaintext, &nonce, &aad,
 				&c->rc.recipient_key));
 
+	if (oscore_msg_inner_decompression != NULL) {
+		if (!oscore_msg_inner_decompression(plaintext->ptr, &plaintext->len))
+			return schc_inner_decomp_error;
+	}
+
 	/* Update nonce only after successful decryption (for handling future responses) */
 	if (NULL != new_nonce_oscore_option) {
 		TRY(byte_array_cpy(&c->rrc.nonce, &nonce, NONCE_LEN));
@@ -325,7 +331,7 @@ decrypt_wrapper(struct byte_array *ciphertext, struct byte_array *plaintext,
 }
 
 enum err oscore2coap(uint8_t *buf_in, uint32_t buf_in_len, uint8_t *buf_out,
-		     uint32_t *buf_out_len, struct context *c)
+		     uint32_t *buf_out_len, struct context *c, plaintext_cb_t oscore_msg_inner_decompression)
 {
 	struct o_coap_packet oscore_packet;
 	struct compressed_oscore_option oscore_option;
@@ -388,7 +394,7 @@ enum err oscore2coap(uint8_t *buf_in, uint32_t buf_in_len, uint8_t *buf_out,
 
 		/* Decrypt packet using new nonce based on the packet */
 		TRY(decrypt_wrapper(ciphertext, &plaintext, c, &oscore_option,
-				    &oscore_packet, &output_coap));
+				    &oscore_packet, &output_coap, oscore_msg_inner_decompression));
 
 		if (ECHO_REBOOT == c->rrc.echo_state_machine) {
 			/* Abort the execution if this is the the first request after reboot.
@@ -438,7 +444,8 @@ enum err oscore2coap(uint8_t *buf_in, uint32_t buf_in_len, uint8_t *buf_out,
 				TRY(decrypt_wrapper(ciphertext, &plaintext, c,
 						    &oscore_option,
 						    &oscore_packet,
-						    &output_coap));
+						    &output_coap,
+						    oscore_msg_inner_decompression));
 
 				/*update replay protection value in context*/
 				TRY(notification_number_update(
@@ -456,12 +463,14 @@ enum err oscore2coap(uint8_t *buf_in, uint32_t buf_in_len, uint8_t *buf_out,
 				TRY(decrypt_wrapper(ciphertext, &plaintext, c,
 						    &oscore_option,
 						    &oscore_packet,
-						    &output_coap));
+						    &output_coap,
+						    oscore_msg_inner_decompression));
 			} else {
 				/*response without PIV*/
 				TRY(decrypt_wrapper(ciphertext, &plaintext, c,
 						    NULL, &oscore_packet,
-						    &output_coap));
+						    &output_coap,
+						    oscore_msg_inner_decompression));
 			}
 		}
 	}
